packages/storage/src/database/common/merkle-tree.ts

```diff
@@ -220,57 +197,51 @@ export class ModelMerkleTree extends ModelGeneral<
     }
 
     witnessPath.push(
-      await this.getNode(this._height - 1, 0n, updatedAt, options).then(
-        (node) => {
-          return {
-            hash: node.toString(),
-            level: this.height - 1,
-            index: Number(0n),
-            witness: false,
-            target: false,
-            empty: node.equals(this.zeroes[this._height - 1]).toBoolean(),
-          };
-        }
-      )
+      await this.getNode(this._height - 1, 0n, before, options).then((node) => {
+        return {
+          hash: node.toString(),
+          level: this.height - 1,
+          index: 0n,
+          witness: false,
+          target: false,
+          empty: node.equals(this.zeroes[this._height - 1]).toBoolean(),
+        };
+      })
     );
     return witnessPath;
   }
 
   public async getNode(
     level: number,
     index: bigint,
-    updatedAt: Date,
+    before: Date,
     options?: FindOptions
   ): Promise<Field> {
-    try {
-      const nodeId = ModelMerkleTree.encodeLevelAndIndexToObjectId(
-        level,
-        index
-      );
-
-      const query = {
-        nodeId,
-        updatedAt: { $lte: updatedAt },
-      };
-
-      const node = await this.collection
-        .find(query, options)
-        .sort({ updatedAt: -1 }) // Gets the latest state at or before the specified updatedAt
-        .limit(1)
-        .toArray();
-
-      if (node.length === 0) {
-        return this.zeroes[level];
-      }
+    const node = await this.collection
+      .find(
+        {
+          $and: [
+            { level },
+            { index },
+            {
+              updatedAt: { $lte: before },
+            },
+          ],
+        },
+        options
+      )
+      .sort({ updatedAt: -1 }) // Gets the latest state at or before the specified updatedAt
+      .limit(1)
```

Comment:
find all document and limit 1? findOne would be better, it will scan to first match filter and return. Instead of scan a whole things and return 1